---
title: "Analysis of nuMt allele frequencies"
author: "Richard Nichols"
date: "21/12/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Pre-processing in the data


### Initial parameter choices 

```{r, echo=TRUE}

# The lowest average frequency the minor allele 
cutoff <- 0.01
# The no. of SNPs to sample (glmer gets indigestion with the full dataset)
snpNo <- 200 


```


### Raw data

The datafile (read from github.com/qmwugbt112/nuMt-analysis),
gives the

  * the proportion of all reads that are mtDNA-like (i.e. are either organellar or nuMt)
  * counts of the 
    + alternative allele (i.e. not found in organellar sequence)
    + main allele 
  * names for 
    + the SNP
    + the individual insect
  * a boolean variable identifying sites thought to be polymorphic in the organellar mitochondrial DNA (which would muddy the analysis). 

```{r , echo=FALSE, cache=TRUE}
temp <- tempfile()
download.file("https://github.com/qmwugbt112/nuMt-analysis/raw/main/mapDepFreq.csv.zip", temp)
binDat <- read.csv(unz(temp,'mapDepFreq.csv'))
rm(temp)
library(knitr)
kable(binDat[1:7,], caption="Raw data")
```



```{r, include=FALSE}
# Some initial data processing creates 3 matrixes
# arranged so all the information from one individual is in a row 
# (& each SNP site is a column):
#   altMat: counts of the rarer allele
#   mainMat: counts of the common allele and another of its alternative)
#   polymat: TRUE values to indicate that the site is thought
#             to be polymorphic in live mitochondria
#             (& hence excluded from this analysis)

altMat <- with(binDat, matrix(alt, 
                              nrow = nlevels(individual),
                              ncol = nlevels(SNP)
                              )
               )

mainMat <- with(binDat, matrix(main, 
                               nrow = nlevels(individual),
                               ncol = nlevels(SNP)
                               )
                )

polyMat <- with(binDat, matrix(polymorphic, 
                            nrow = nlevels(individual),
                            ncol = nlevels(SNP))
                )

# Sum the columns of the numerical matixes
altsum <- colSums(altMat)
mainsum <- colSums(mainMat)

# Identify the columns of TRUE values in polyMat 
# (which indicute the SNP sites to be excluded)
polysum <- apply(polyMat,2,unique)

```

### Initial filtering
These three plots show

* The distribution of raw allele frequency data for each SNP across all individuals
* The truncated data (Frequencies > 1%)
* The subsampled data (SNPs selected in proportion to their frequency)

The idea behind the subsampling method is that the higher-frequency alleles are more informative about the proportion of the nuclear genome that is nuMt DNA, so they more heavily sampled. However we wish to cover the whole frequency spectruum; in case rare alleles show different trends, for example.

```{r, echo=FALSE}
par(mfrow = c(3,1))
par(mar=c(2,1,1,1))

# investigate shape of the frequency distribution all data
freq <- altsum/(altsum+mainsum)
h <- hist(freq,
       breaks = 0:500 / 500 * 0.4, 
       main = '', 
       xlab = 'Frequency')
text(x = 0.15, 
     y= max(h$counts)*0.6,
     paste('Raw allele frequencies. N=',sum(h$counts),'SNPs')
     )

# repeat for truncated data
h <- hist(freq[(!polysum) & (freq > cutoff)],
          breaks = 0:500 / 500 * 0.4,
          main = '',
          xlab = 'Frequency')
text(x = 0.15,
     y= max(h$counts)*0.6,
     paste('Truncated allele frequencies. N=',sum(h$counts),'SNPs')
     )

# calculate boolean vector to dataframe to subsample SNPs for analysis
# only choose sites without mt polymorphism and relatively frequent
choice1 <- (!polysum) & (freq > cutoff)

# we will be weighting sampling probability towards high freq loci
# and exclude those we have ruled out above by setting their freq2 to zero
freq2 <- freq
freq2[!choice1] <- 0



# Choose 200 loci at random weighting by allele frequency
choice2 <- sample(1:length(freq),
                  size = snpNo,
                  prob = freq2)
h <- hist(freq[choice2],
          breaks = 0:500 / 500 * 0.4,
          main = '',
          xlab = 'Frequency')
text(x = 0.15,
     y= max(h$counts)*0.6,
     paste('Subsampled SNPs. N=',sum(h$counts),'SNPs')
     )

# reset the graphics window
par(mfrow = c(1,1))

```

### Viewing the subsampled data

```{r, echo=FALSE}
# Create a boolean vector to record T for the selected SNPs F for the others
chosen <- rep(FALSE, length(freq))
chosen[choice2] <- TRUE

# repeat each element by the number of individuals so that the new
# vector can be used to select records from the full dataset.
chosen <- rep(chosen, each = nlevels(binDat$individual))

# Create a dataframe with just the subsampled data
subDat <- subset(binDat, chosen)

# read the factor on top of itself to reduce the levels in SNP to 200
subDat$SNP <- factor(subDat$SNP)

# cleanup - remove all but dataframes and snpNo
listing <- ls()
rm(list = listing[ -c(
  grep("Dat",listing), 
  grep("No",listing))])



# calculate x & y values for the regression
# the -qlogis function converts mapped reads / (all reads)
# to log(unmapped reads / mapped reads)
subDat$xvals <- with(subDat, -qlogis(mapDep))
subDat$yvals <- with(subDat, log(alt/(alt+main)))

# identify the older data using different sequencing technology
subDat$older <- with(subDat, factor(
  is.element(individual, levels(individual)[47:52]))
  )

# remove allele frequencies of zero (-Inf log)
subDat$yvals[subDat$yvals == -Inf] <- NA

# construct a regression through the orgin
# with slope 1 (offset = xvals)
mod1 <- lm(yvals ~ 0 + older + SNP,
           offset = xvals,
           data = subDat)

# store the results needed for plotting
cvals <- coef(mod1)
minval <- min(cvals[-(1:2)] + cvals[1])
maxval <- max(cvals[-(1:2)] + cvals[1])

par(mar = c(4,4,3,1))

# plot the data and regression
with(subDat,
     plot(xvals, yvals,
          main = 'Raw frequency data\n(older data in blue)',
          xlab = '(Unmapped reads / Mapped)',
          ylab = 'Allele Frequency (log scale)',
          col = rainbow(2)[as.numeric(older)],
          cex = 0.2,
          xlim = c(0, max(xvals) * 1.01),
          ylim = c(minval, max(c(0,yvals), na.rm = T) * 1.1),
          xaxt = 'n',
          yaxt = 'n'
          )
     
     )

axis(1, at = log(2^(0:3*3)), labels = 2^(0:3*3)) 
ll <- expression("0", "10"^-1, "10"^-2,"10"^-3,"10"^-4,"10"^-5,"10"^-6)
axis(2, at = log(10^(0:-6)), labels = ll)

abline(minval, 1)
abline(maxval, 1)
abline(h = -max(subDat$xvals), lty = 3)
abline(v = 0, lty = 3)


```


```{r}
summary(mod1)
```



This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.
